File: HackerRank/InterviewPreparation/HashTables/FrequencyQuerries.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the freqQuery function below.
from collections import Counter


def freqQuery(queries):
    arr = []
    result = []
    freq = Counter()
    count = Counter()
    for i in queries:
        if i[0] == 1:
            arr.append(i[1])
            freq[i[1]] += 1
        if i[0] == 2 and i[1] in arr:
            arr.remove(i[1])
        if i[0] == 3:
            if arr.count(i[1]) == 0:
                result.append(0)
            else:
                result.append(1)
    return result


if __name__ == '__main__':
    # fptr = open(os.environ['OUTPUT_PATH'], 'w')

    q = int(input().strip())

    queries = []

    for _ in range(q):
        queries.append(list(map(int, input().rstrip().split())))

    ans = freqQuery(queries)
    print(ans)
    # fptr.write('\n'.join(map(str, ans)))
    # fptr.write('\n')

    # fptr.close()

File: HackerRank/InterviewPreparation/HashTables/FrequencyQuerries.js
function freqQuery(queries) {
  let counts = [];
  let arr = [];
  let freq = {};
  for (let [action, value] of queries) {
    let init = freq[value] || 0;
    if (action == 1) {
      freq[value] = init + 1;
      counts[init] = (counts[init] || 0) - 1;
      counts[init + 1] = (counts[init + 1] || 0) + 1;
    }
    if (action == 2 && init > 0) {
      freq[value] = init - 1;
      counts[init] -= 1;
      counts[init - 1] += 1;
    }
    if (action == 3) {
      arr.push(counts[value] > 0 ? 1 : 0);
    }
  }
  return arr;
}

let queries = [
  [1, 5],
  [1, 6],
  [3, 2],
  [1, 10],
  [1, 10],
  [1, 6],
  [2, 5],
  [3, 2]
];
// queries = [
//   [1, 3],
//   [1, 38],
//   [2, 1],
//   [1, 16],
//   [2, 1],
//   [2, 2],
//   [1, 64],
//   [1, 84],
//   [3, 1],
//   [1, 100],
//   [1, 10],
//   [2, 2],
//   [2, 1],
//   [1, 67],
//   [2, 2],
//   [3, 1],
//   [1, 99],
//   [1, 32],
//   [1, 58],
//   [3, 2]
// ];
let res = freqQuery(queries);
console.log(res);
